glean-core/tests/counter.rs

```diff
@@ -0,0 +1,152 @@
+            ::serde_json::to_string(&snapshot).unwrap(),
+        );
+    }
+
+    // Make a new glean instance here, which should force reloading of the data from disk
+    // so we can ensure it persisted, because it has User lifetime
+    {
+        let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID).unwrap();
+        let snapshot = StorageManager
+            .snapshot_as_json(glean.storage(), "store1".into(), true)
+            .unwrap();
+        assert_eq!(
+            "{\"counter\":{\"telemetry.counter_metric\":1}}",
+            ::serde_json::to_string(&snapshot).unwrap(),
+        );
+    }
+}
+
+#[test]
+fn set_value_properly_sets_the_value_in_all_stores() {
+    let (_t, tmpname) = tempdir();
+    let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID).unwrap();
+    let store_names: Vec<String> = vec!["store1".into(), "store2".into()];
+
+    let metric = CounterMetric::new(CommonMetricData {
+        name: "counter_metric".into(),
+        category: "telemetry".into(),
+        send_in_pings: store_names.clone(),
+        disabled: false,
+        lifetime: Lifetime::Ping,
+        ..Default::default()
+    });
+
+    metric.add(&glean, 1);
+
+    for store_name in store_names {
+        let snapshot = StorageManager
+            .snapshot_as_json(glean.storage(), &store_name, true)
+            .unwrap();
+
+        assert_eq!(1, snapshot.as_object().unwrap().len());
+        println!("{:?}", snapshot);
+        assert_eq!(1, snapshot["counter"]["telemetry.counter_metric"]);
+    }
+}
+
+#[test]
+fn snapshot_returns_none_if_nothing_is_recorded_in_the_store() {
+    let (_t, tmpname) = tempdir();
+    let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID).unwrap();
+    assert!(StorageManager
+        .snapshot(glean.storage(), "unknown_store", true)
+        .is_none())
+}
+
+#[test]
+fn snapshot_correctly_clears_the_stores() {
+    let (_t, tmpname) = tempdir();
+    let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID).unwrap();
+    let store_names: Vec<String> = vec!["store1".into(), "store2".into()];
+
+    let metric = CounterMetric::new(CommonMetricData {
+        name: "counter_metric".into(),
+        category: "telemetry".into(),
+        send_in_pings: store_names.clone(),
+        disabled: false,
+        lifetime: Lifetime::Ping,
+        ..Default::default()
+    });
+
+    metric.add(&glean, 1);
+
+    // Get the snapshot from "store1" and clear it.
+    let snapshot = StorageManager.snapshot(glean.storage(), "store1", true);
+    assert!(snapshot.is_some());
+    // Check that getting a new snapshot for "store1" returns an empty store.
+    assert!(StorageManager
+        .snapshot(glean.storage(), "store1", false)
+        .is_none());
+    // Check that we get the right data from both the stores. Clearing "store1" must
+    // not clear "store2" as well.
+    let snapshot2 = StorageManager.snapshot(glean.storage(), "store2", true);
+    assert!(snapshot.unwrap().len() != 0);
```

Comment:
Instead of checking the length, we can just check that it `is_some`, just as above.
Also checking `snapshot` again seems weird, that check was done in line 111.

Reply:
It's not quite checking the same thing, though.  The function could easily return `Some(Object::new())` which would pass `is_some()`, but fail `unwrap().len() != 0`.  I think it's important to assert that *if* it returns `Some` that `Object` also has at least one entry.

Reordering things makes sense, though.

Reply:
Hm.... `snapshot` already has this check though and it serializes a HashMap. Even an empty hashmap is at least serialized as `{}`, therefore defeating your test.

IMO it would be much more sensible to test that we get `None` when we expect it and `Some` to indicate it contains data and to look into the data if needed.

(And at least `len() != 0` should be written as `!...is_empty()`)

Reply:
Got it -- I forgot that `snapshot` is snapshotting as a string.